### DISCUSSION TITLE
Using custom typescript express server

### DISCUSSION THREAD

**Discussion by spy16** [remix-run/remix#6004] [Q&A]
Hi, I am trying use custom express server for my app. When the app is created with the express template, it creates a `server.js`. Rest of my project is written in TS and i want to be able to import some of those modules in my server file...

My question is how do I turn `server.js` into `server.ts`? 

Things I have tried:

1. I tried to follow the structure of [this repo](https://github.com/kiliman/remix-ras-server-example/blob/main/package.json) by @kiliman. But it uses old version of remix with `serverEntryFile` config which doesn't seem to exist anymore. Doing everything else works when I build the project but in dev mode, `remix dev` still falls back to remix-app-server.
2. I tried to set `server` config to `./server.ts` and put `createApp(...).listen(...)` inside that. But with that `remix dev` gives this error: `Error: remix dev is not supported for custom servers.` 

Please help 🙏. I am stuck. All I want to do is run some logic before the server starts.

**Comment by kiliman** [comment]
That repo was an experimental version that never made it into Remix.

The thing is that Remix doesn't care about your server file. You run that separately from `remix watch` (the build service).

The simplest is to add `ts-node` to your dev dependencies. Then change your *package.json* *server.js* to *server.ts*

```json
"dev:node": ".... ./server.ts"
```
`nodemon` will run `ts-node` when it sees a TypeScript file.

**Reply by spy16** [reply]
Yes. I tried this. 

* changed `server.js` references in the `package.json` to `server.ts`.
* Added `ts-node` so that nodemon can use it for ts file. Also added all the missing `@types/*`
* Also changed my build to `remix build && esbuild --format=cjs --platform=node --outfile=build/server.js server.ts` 

Both build and dev steps worked perfectly. I could even run `node build/server.js` and everything would work as expected. Then I imported `~/services/data.server.ts` in my `server.ts` - Running `yarn dev` or `yarn build` still seemed to work. But if running `node build/server.js` after building in this config, I get this:

```
Error: Cannot find module '~/services/data.server'
Require stack:
``` 

In the esbuild output `build/server.js` it still has `~/services/data.server` import itself. i don't know how to make this part of the remix build itself so that all these dependency modules get referenced correctly.

**Reply by cliffordfajardo** [reply]
@spy16 - @xHomu has created a guide for the new v2 dev server that has a `server.ts` example

https://github.com/xHomu/remix-v2-server
I discovered this via the https://remix.guide/ website ❤️ 
<img width="400" alt="image" src="https://github.com/remix-run/remix/assets/6743796/99b817a8-9930-4ca5-8dfa-d97a6877dee8">

**Comment by koderyoda** [comment]
Getting server.js to place nice with the rest of remix is imo a pretty basic need. I followed the tutorial to set up socketio with remix and it worked except I could not import any ts module (all my remix app is ts) into server.js. Long story short, one of the answers in this [SO thread](https://stackoverflow.com/questions/62096269/cant-run-my-node-js-typescript-project-typeerror-err-unknown-file-extension) referred to tsx and it worked like a charm. Now my dev script reads like:
`"dev": "remix dev --manual -c \"tsx -r dotenv/config --watch server.js\"",`

**Reply by RoniqueRicketts** [reply]
I am still exeperiencing this issue with Vite I am running remix-run v 2.8.1 It'd be nice to have and use TS on my the backend.

**Comment by vizardkill** [comment]
UP

**Reply by kiliman** [reply]
I commented here on how you can not only use Typescript, but also have a single build for your Express and Remix app.

https://github.com/remix-run/remix/discussions/5783#discussioncomment-10627308

**Reply by gustavopch** [reply]
> but also have a single build for your Express and Remix app.

Gonna leave this alternative workaround here too for those who don't need middleware and server context: https://github.com/remix-run/remix/discussions/9790#discussioncomment-10625633

**Reply by mihaiandrei97** [reply]
I also created a repo that allows you to use server.ts and builds the server with esbuild (check vite.config). In dev it is using tsx

https://github.com/mihaiandrei97/remix-express

**Reply by gustavopch** [reply]
@mihaiandrei97 Yeah, that's how I do (I actually use `tsx` in prod as well to run my `server.ts`). Just beware that if you import anything from the `app` folder, you'll get a different instance compared to the exact same import happening elsewhere in your project, which is the problem that's described here: https://github.com/remix-run/remix/discussions/9790. If you want to avoid that problem, either use `remix-express-vite-plugin` or do as I described in the link in my previous comment.

**Reply by mihaiandrei97** [reply]
@gustavopch thank you for pointing it out! I was not aware of it. In my case, I think it is ok since I just need to pass the user to the context. But good to know 🙏
